src/memory_manager.rs

```diff
@@ -287,13 +397,96 @@ impl<M: Memory> MemoryManagerInner<M> {
             bucket_size_in_pages: header.bucket_size_in_pages,
             memory_sizes_in_pages: header.memory_sizes_in_pages,
             memory_buckets,
+            freed_buckets: BTreeSet::new(),
+        }
+    }
+
+    fn load_layout_v2(memory: M, header: Header) -> Self {
+        let mut memory_size_in_buckets = vec![];
+        let mut number_of_used_buckets = 0;
+
+        for memory_size_in_pages in header.memory_sizes_in_pages.into_iter() {
+            let size_in_buckets = (memory_size_in_pages + header.bucket_size_in_pages as u64 - 1)
+                / header.bucket_size_in_pages as u64;
+            memory_size_in_buckets.push(size_in_buckets);
+            number_of_used_buckets += size_in_buckets;
+        }
+
+        const BYTE_SIZE_IN_BITS: usize = 8;
+        let buckets_index_size_in_bytes: usize =
+            (number_of_used_buckets as usize * BUCKET_ID_LEN_IN_BITS + (BYTE_SIZE_IN_BITS - 1))
+                / BYTE_SIZE_IN_BITS;
+
+        let mut buckets = vec![0; buckets_index_size_in_bytes];
+        memory.read(bucket_indexes_offset().get(), &mut buckets);
+
+        let buckets_decompressed = bytes_to_bucket_indexes(&buckets);
+
+        let mut memory_buckets = BTreeMap::new();
+
+        let mut last_occupied_bucket: i32 = -1;
+
+        let mut bucket_idx: usize = 0;
+
+        for (memory, size_in_buckets) in memory_size_in_buckets.into_iter().enumerate() {
+            let mut vec_buckets = vec![];
+            for _ in 0..size_in_buckets {
+                let bucket = buckets_decompressed[bucket_idx];
+                last_occupied_bucket = std::cmp::max(bucket.0 as i32, last_occupied_bucket);
+                vec_buckets.push(bucket);
+                bucket_idx += 1;
+            }
+            memory_buckets
+                .entry(MemoryId(memory as u8))
+                .or_insert(vec_buckets);
+        }
+
+        if last_occupied_bucket == -1 {
+            Self {
+                memory,
+                allocated_buckets: header.num_allocated_buckets,
+                bucket_size_in_pages: header.bucket_size_in_pages,
+                memory_sizes_in_pages: header.memory_sizes_in_pages,
+                memory_buckets,
+                freed_buckets: BTreeSet::new(),
+            }
+        } else {
+            let mut freed_buckets = BTreeSet::new();
+
+            for id in 0..last_occupied_bucket as u16 {
+                freed_buckets.insert(BucketId(id));
+            }
+
+            for id in buckets_decompressed.iter() {
+                freed_buckets.remove(id);
+            }
+
+            Self {
+                memory,
+                allocated_buckets: header.num_allocated_buckets,
+                bucket_size_in_pages: header.bucket_size_in_pages,
+                memory_sizes_in_pages: header.memory_sizes_in_pages,
+                memory_buckets,
+                freed_buckets,
+            }
+        }
```

Comment:
Generally speaking, try to reduce if statements in your code. It results in more readable and maintainable code. Here's how I'd replace this if statement:


```
        let mut freed_buckets: BTreeSet<BucketId> = (0..max_bucket_id).map(BucketId).collect();

        for id in buckets.iter() {
            freed_buckets.remove(id);
        }

        Self {
            memory,
            allocated_buckets: header.num_allocated_buckets,
            bucket_size_in_pages: header.bucket_size_in_pages,
            memory_sizes_in_pages: header.memory_sizes_in_pages,
            memory_buckets,
            freed_buckets,
        }
```

Reply:
Good point, done.